code-styling/bash/README.md

```diff
@@ -0,0 +1,423 @@
+
+- Error messages MUST be redirected to STDERR. 
+  This makes it easier to separate normal status from actual issues.)
+
+- The following exit codes MUST NOT be used as they already have a specific 
+  meaning:
+  
+  - `1` Catchall for general errors
+  - `2` Misuse of shell builtins
+  - `126` Invoked command is not executable
+  - `127` "command not found"
+  - `128` Invalid argument to exit
+  - `128+n` Fatal error signal "`n`"
+  - `130` Script terminated by <kbd>Control</kbd>-<kbd>C</kbd>
+  - `255*` Numbers above `255` loop round from `0`
+
+- To distinguish them from system errors, exit codes other than `0` SHOULD start
+  from `64`
+
+- Exit code `64` is RECOMMENDED to be used as generic/unspecified error code.
+
+<!-- ======================================================================= -->
+
+## Formatting
+
+### Line length
+
+- Lines SHOULD be less than 80 characters and MUST be less than 120 characters;
+  longer lines SHOULD be split into multiple subsequent lines.
+
+### Whitespace 
+
+#### Lines
+
+- Lines MUST use an indentation of 4 spaces and MUST NOT use tabs for indenting.
+
+- There SHOULD NOT be trailing whitespace at the end of any line.
+
+- Consecutive blank lines MAY be added to improve readability and to indicate 
+  related blocks of code.
+
+- If a pipeline fits on one line, it SHOULD be on one line. To improve
+  readability it MAY be split across several lines.
+
+- If a pipeline does not fit on one line it MUST be split across several lines.
+
+- When splitting, a line MUST be indented and begin with a pipe '|' or logical 
+  compound ('||' and '&&').  Example:
+  
+  ```bash
+  command \
+      | pipe \
+      && other
+  ```
+
+- The `; do` and `; then` keywords MUST be on the same line as the `while`, `for` 
+  or `if` it belongs to.
+
+<!-- ======================================================================= -->
+
+## Declaration 
+
+### Arrays
+
+- For arrays, either use `"$@"` or `$*` MUST be used .
+  An unquoted `$@` MUST NOT be used.
+
+### Checks
+
+- The new BASH double square bracket builtin test operator (`[[ ... ]]`) MUST be
+  used. The old single square bracket test operator `[` or explicit call
+  to `test` MUST NOT be used.
+
+- To avoid confusion about what is being tested for `-z` or `-n` MUST be used.
+  Use `[[ -z $variable ]]`  or `[[ -n $variable ]]` rather than just `[[ $variable ]]`.
+
+- A single equal sign SHOULD be used when checking values. Double or triple 
+  signs SHOULD NOT be used.
+  Example: `[[ "${NAME}" = "John" ]]` 
+
+- When referring to the current file `${BASH_SOURCE[0]}` SHOULD be used, even if 
+  it is sourced from a parent script. In other cases, `${0}` SHOULD be used.
+
+- As the `errexit` flag is used, `|| true` MUST be appended to commands that 
+  might cause an error which should not halt the script. (For instance when the
+  command output is assigned to a variable).
+
+### Command options
+
+- Long options MUST be used if they are available, 
```

Comment:
This could cause problems with all platforms or incompatibility. (for example the bash version on Mac OS X)